drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c

```diff
@@ -0,0 +1,1669 @@
+	} else {
+		struct bcm2835_isp_q_data *q_data = &node->q_data;
+
+		f->fmt.pix.width = q_data->width;
+		f->fmt.pix.height = q_data->height;
+		f->fmt.pix.field = V4L2_FIELD_NONE;
+		f->fmt.pix.pixelformat = q_data->fmt->fourcc;
+		f->fmt.pix.bytesperline = q_data->bytesperline;
+		f->fmt.pix.sizeimage = q_data->sizeimage;
+		f->fmt.pix.colorspace = q_data->fmt->colorspace;
+	}
+
+	return 0;
+}
+
+static int bcm2835_isp_node_enum_fmt(struct file *file, void  *priv,
+				     struct v4l2_fmtdesc *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt_list *fmts = &node->supported_fmts;
+
+	if (f->type != node->queue.type)
+		return -EINVAL;
+
+	if (f->index < fmts->num_entries) {
+		/* Format found */
+		f->pixelformat = fmts->list[f->index].fourcc;
+		f->flags = fmts->list[f->index].flags;
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+static int bcm2835_isp_node_try_fmt(struct file *file, void *priv,
+				    struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt *fmt;
+
+	if (f->type != node->queue.type)
+		return -EINVAL;
+
+	if (!node_is_stats(node)) {
+		f->fmt.pix.width = max(min(f->fmt.pix.width, MAX_DIM),
+				       MIN_DIM);
+		f->fmt.pix.height = max(min(f->fmt.pix.height, MAX_DIM),
+					MIN_DIM);
+
+		fmt = find_format(f, node);
+		if (!fmt) {
+			f->fmt.pix.pixelformat =
+					get_default_format(node)->fourcc;
+			fmt = find_format(f, node);
+		}
+
+		f->fmt.pix.colorspace = fmt->colorspace;
+		f->fmt.pix.bytesperline = get_bytesperline(f->fmt.pix.width,
+							   fmt);
+		f->fmt.pix.field = V4L2_FIELD_NONE;
+		f->fmt.pix.sizeimage =
+			get_sizeimage(f->fmt.pix.bytesperline, f->fmt.pix.width,
+				      f->fmt.pix.height, fmt);
+	} else {
+		fmt = find_format(f, node);
+		if (!fmt) {
+			f->fmt.meta.dataformat =
+					V4L2_META_FMT_BCM2835_ISP_STATS;
+			fmt = find_format(f, node);
+		}
+		f->fmt.meta.buffersize =
+				get_port_data(node)->minimum_buffer.size;
+	}
+
+	return 0;
+}
+
+static int bcm2835_isp_node_s_fmt(struct file *file, void *priv,
+				  struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	int ret;
+
+	if (f->type != node->queue.type)
+		return -EINVAL;
+
+	ret = bcm2835_isp_node_try_fmt(file, priv, f);
+	if (ret)
+		return ret;
+
+	v4l2_dbg(1, debug, &node_get_dev(node)->v4l2_dev,
+		 "%s: Set format for node %s[%d]\n",
+		 __func__, node->name, node->id);
+
+	return populate_qdata_fmt(f, node);
+}
+
+static int bcm2835_isp_node_s_selection(struct file *file, void *fh,
+					struct v4l2_selection *s)
+{
+	struct mmal_parameter_crop_t crop;
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_dev *dev = node_get_dev(node);
+	struct vchiq_mmal_port *port = get_port_data(node);
+
+	/* This return value is required fro V4L2 compliance. */
+	if (node_is_stats(node))
+		return -ENOTTY;
+
+	if (s->type != V4L2_BUF_TYPE_VIDEO_OUTPUT)
+		return -EINVAL;
```

Comment:
This feels odd after a check of the node type.

Does this mean that setting a type != V4L2_BUF_TYPE_VIDEO_OUTPUT on either OUTPUT or CAPTURE nodes gets accepted? I'd expect only setting with type = V4L2_BUF_TYPE_VIDEO_OUTPUT && on the OUTPUT node would be required.

Reply:
This was down to the compliance failing.  However, with the relevant ioctls disabled, it's not needed any more.

Reply:
Actually, it's still a problem with compliance that forces an invalid value into s->type.  I have to keep the clause in there.

Reply:
It felt odd rather than being definitely wrong. I'd half suggest ```if (s->type != node->type)``` again, but I don't see it ever getting enabled on the CAPTURE queues.

Reply:
Sigh, I think I understand what happened.  Summary is that I need to disable the old crop API ioctls explicitly, otherwise v4l2-compliance (and the v4l2 framework) does some strange things and goes into the s/g_selection function **even though they have been disabled**.